### PR TITLE
[REG-2063] Enable running unity with -rgsequencepath startup option

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Adds -rgsequencepath command line option
  - This allows you to start Unity , automatically run an RG bot sequence, and then exit the unity runtime
  - Also supports -rgsequencetimeout option to put a number of seconds limit on your runtime before it closes Unity and reports a failure
  - Supports RC=0 for successful run and RC=1 for timeout failure and RC=# for various arg parsing errors
- Fixes some cases for -batchmode where keyboard and mouse devices needed to handle being null, and/or needed a virtual keyboard device registered to function correctly 

Usage samples

- Run the game with the full on screen graphics
  - Supports ALL of our SDK features
  - `./com.unity.multiplayer.samples.coop.exe -rgsequencepath "Assets/RegressionGames/Resources/BotSequences/TypeADF.json" -rgsequencetimeout 7`

- Run the game without rendering
  - Does NOT support anything involving graphics including (screenshots, anything CV related)
  - `./com.unity.multiplayer.samples.coop.exe -batchmode -logFile C:/Users/zack/bossroom_build/headless.log -rgsequencepath "Assets/RegressionGames/Resources/BotSequences/TypeADF.json"`

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        internal static readonly int Rc_SequenceTimeoutNotInt = 7;
+
+        internal static readonly string SequencePathArgument = "-rgsequencepath";
+        internal static readonly string SequenceTimeoutArgument = "-rgsequencetimeout";
+
+        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.BeforeSplashScreen)]
+        public static void FirstLoadChecks()
+        {
+            // do this here to fail fast on bad args
+            ParsePathArgument();
```
Comment (batu) [inline src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs:34]: Good stuff.

File: src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.AfterSceneLoad)]
+        public static void Initialize()
+        {
+            var resourcePath = ParsePathArgument();
+            if (resourcePath != null)
+            {
+                BotSegmentsPlaybackController playbackController = Object.FindObjectOfType<BotSegmentsPlaybackController>();
+                if (playbackController == null)
+                {
+                    RGDebug.LogError($"RGOverlay must be in the first scene loaded for your game to utilize {SequencePathArgument} command line argument");
```
Comment (batu) [inline src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs:47]: Do we see this in the console when running in headless mode? If not, I think we should definitely make a note in docs to point people to the correct log file to debug sudden closure of their app

File: src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                    Application.Quit(Rc_OverlayNotInScene);
+                }
+
+                playbackController.gameObject.AddComponent<RGHeadlessSequenceRunnerBehaviour>();
+
+                //ensure a keyboard exists for the input playback
+                if (IsBatchMode())
+                {
+                    var currentKeyboard = Keyboard.current;
```
Comment (batu) [inline src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs:56]: If they already have a keyboard would we also need to attempt to set flags?

**Review by batu — APPROVED** [review]
This looks good! Added a few questions for my comprehension

File: src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs (hunk 3 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs:56]: We definitely do not today as we don't create a virtual keyboard for our runs outside of this headless environment.

It's a good philosophical discussion we should experiment with though in the future

File: src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs (hunk 2 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/RGHeadlessSequenceRunner.cs:47]: In console mode you get no STDOUT from Unity.  Everything is put into the -logfile path you specify, or defaults to the player.log file if not running in -batchmode

That said, I'm not very interested in re-documenting how Unity works for users 👎 

We absolutely should/will document that this feature exists once we've had a cycle or two of testing on it, 
https://linear.app/regression-games/issue/REG-2074/document-rgsequencepath-command-line-options

**Review by vontell — APPROVED** [review]
Nice! No additional comments from my side. I wonder if we want to add some tests, although in this case I don't really mind since this was mostly just to make sure things can work in headless mode, I imagine the testing around the remote execution of a sequence will be time better spent.
